runtime/jailer.go

```diff
@@ -0,0 +1,403 @@
+			}
+
+			contentsPath := j.ContentsPath()
+			j.logger.WithField("contentsPath", contentsPath).Debug("Creating firecracker contents path")
+			if err := os.MkdirAll(contentsPath, mode); err != nil {
+				return errors.Wrapf(err, "failed to create contents path: %v", contentsPath)
+			}
+
+			rootPathToConfig := filepath.Join(ociBundlePath, "config.json")
+			j.logger.WithField("rootPathToConfig", rootPathToConfig).Debug("Copying config")
+			if err := copyFile(runcConfigPath, rootPathToConfig, 0666); err != nil {
+				return errors.Wrapf(err, "failed to copy config from %v to %v", runcConfigPath, rootPathToConfig)
+			}
+
+			j.logger.Debug("Overwritting process args of config")
+			if err := j.overwriteConfig(cfg, filepath.Base(m.Cfg.SocketPath), rootPathToConfig); err != nil {
+				return errors.Wrap(err, "failed to overwrite config.json")
+			}
+
+			// copy the firecracker binary
+			j.logger.WithField("binPath", binPath).Debug("copying firecracker binary")
+			if err := copyFile(cfg.FirecrackerBinaryPath, binPath, 0700); err != nil {
+				return errors.Wrapf(err, "could not copy firecracker binary from path %v", cfg.FirecrackerBinaryPath)
+			}
+
+			// Link the kernel image
+			newKernelImagePath := filepath.Join(contentsPath, kernelImageFileName)
+			j.logger.WithField("newKernelImagePath", newKernelImagePath).Debug("linking kernel image")
+			if err := os.Link(m.Cfg.KernelImagePath, newKernelImagePath); err != nil {
+				return errors.Wrapf(err, "failed to link kernel image from path %v", m.Cfg.KernelImagePath)
+			}
+			m.Cfg.KernelImagePath = filepath.Join("/var/lib/firecracker-containerd", kernelImageFileName)
+
+			// link drives to new contents path
+			for i, d := range m.Cfg.Drives {
+				drivePath := firecracker.StringValue(d.PathOnHost)
+				fileName := filepath.Base(drivePath)
+
+				if err := os.Link(drivePath, filepath.Join(contentsPath, fileName)); err != nil {
+					return errors.Wrapf(err, "failed to copy drive %v", drivePath)
+				}
+
+				newDrivePath := filepath.Join(strings.TrimPrefix(contentsPath, rootPath), fileName)
+				m.Cfg.Drives[i].PathOnHost = firecracker.String(newDrivePath)
+			}
+
+			if err := chownr(
+				rootPath,
+				int(j.Config.UID),
+				int(j.Config.GID),
+			); err != nil {
+				return errors.Wrapf(err, "failed to chown -r %v", rootPath)
+			}
+
+			j.logger.Info("Successfully ran jailer handler")
+			return nil
+		},
+	}
+}
+
+// BuildLinkFifoHandler will return a new firecracker.Handler with the function
+// that will allow linking of the fifos making them visible to Firecracker.
+func (j runcJailer) BuildLinkFifoHandler() firecracker.Handler {
+	return firecracker.Handler{
+		Name: jailerFifoHandlerName,
+		Fn: func(ctx context.Context, m *firecracker.Machine) error {
+			contentsPath := j.ContentsPath()
+			fifoFileName := filepath.Base(m.Cfg.LogFifo)
+			newFifoPath := filepath.Join(contentsPath, fifoFileName)
+			if err := os.Link(m.Cfg.LogFifo, newFifoPath); err != nil {
```

Comment:
Do we actually need the FIFOs to be available both inside an outside the jail? If they only exist in the jail (possibly created by `BuildHandler()`), they're still reachable by clients outside the jail.

Reply:
Technically it should only be needed for the inside. The reason is due to Firecracker having only visibility of the jailed rootfs. However, if the user is doing something with those fifos on their end, having them outside would be necessary.